x-pack/plugins/security/server/audit/audit_events.ts

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { KibanaRequest } from 'src/core/server';
+import { AuthenticationResult } from '../authentication/authentication_result';
+
+/**
+ * Audit event schema using ECS format.
+ * https://www.elastic.co/guide/en/ecs/1.5/index.html
+ * @public
+ */
+export interface AuditEvent {
+  /**
+   * Human readable message describing action, outcome and user.
+   *
+   * @example
+   * User [jdoe] logged in using basic provider [name=basic1]
+   */
+  message: string;
+  event: {
+    action: string;
+    category?: EventCategory;
+    type?: EventType;
+    outcome?: EventOutcome;
+    module?: string;
+    dataset?: string;
+  };
+  user?: {
+    name: string;
+    email?: string;
+    full_name?: string;
+    hash?: string;
+    roles?: readonly string[];
+  };
+  kibana?: {
+    /**
+     * Current space id of the request.
+     */
+    space_id?: string;
+    /**
+     * Saved object that was created, changed, deleted or accessed as part of the action.
+     */
+    saved_object?: {
+      type: string;
+      id?: string;
+    };
+    /**
+     * Any additional event specific fields.
+     */
+    [x: string]: any;
+  };
+  error?: {
+    code?: string;
+    message?: string;
+  };
+  http?: {
+    request?: {
+      method?: string;
+      body?: {
+        content: string;
+      };
+    };
+    response?: {
+      status_code?: number;
+    };
+  };
+  url?: {
+    domain?: string;
+    full?: string;
+    path?: string;
+    port?: number;
+    query?: string;
+    scheme?: string;
+  };
+}
+
+export enum EventCategory {
+  DATABASE = 'database',
+  WEB = 'web',
+  IAM = 'iam',
+  AUTHENTICATION = 'authentication',
+  PROCESS = 'process',
+}
+
+export enum EventType {
+  USER = 'user',
+  GROUP = 'group',
+  CREATION = 'creation',
+  ACCESS = 'access',
+  CHANGE = 'change',
+  DELETION = 'deletion',
+}
+
+export enum EventOutcome {
+  SUCCESS = 'success',
+  FAILURE = 'failure',
+  UNKNOWN = 'unknown',
+}
+
+export interface HttpRequestParams {
+  request: KibanaRequest;
+}
+
+export function httpRequestEvent({ request }: HttpRequestParams): AuditEvent {
+  const [path, query] = request.url.path?.split('?') ?? [];
```

Comment:
I somehow submitted my review without including this, whoops!

As discussed offline, we need to include the space ID prefix in the path.

```suggestion
export function httpRequestEvent(
  { request }: HttpRequestParams,
  spaceId: string | undefined
): AuditEvent {
  const [pathSuffix, query] = request.url.path?.split('?') ?? [];
  // If this request is in the context of a non-Default space, the spaceId prefix portion of the path was stripped out by the Spaces plugin.
  // We need to recreate the original path to ensure that Kibana audit events can be correlated with those from external systems.
  // Note: this approach will need to change after https://github.com/elastic/kibana/issues/62053 is implemented.
  const path = spaceId ? `/s/${spaceId}${pathSuffix}` : pathSuffix;
```

Reply:
My concern with manually reverting the rewrite performed by the spaces plugin here is that other plugins could rewrite urls as well and we'd have no way of knowing about it which would be unreliable. 

I've discussed it with @restrry and the platform team will add a way to read out the original url from the request object for us similar to expressjs's `req.originalUrl` property.

Reply:
#80485